compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -4324,6 +4328,139 @@ public void visit(BLangStatementExpression bLangStatementExpression) {
+    public void visit(BLangQueryExpr queryExpr) {
+        List<BLangFromClause> fromClauseList = queryExpr.fromClauseList;
+        BLangFromClause fromClause = fromClauseList.get(0);
+        BLangSelectClause selectClause = queryExpr.selectClause;
+        List<BLangWhereClause> whereClauseList = queryExpr.whereClauseList;
+        DiagnosticPos pos = fromClause.pos;
+
+        // Create Foreach statement
+        //
+        // Below query expression :
+        //      from var person in personList
+        //
+        // changes as,
+        //      foreach var person in personList {
+        //          ....
+        //      }
+
+        BLangForeach leafForEach = null;
+        BLangForeach parentForEach = null;
+
+        for (BLangFromClause bLangFromClause : fromClauseList) {
+            BLangForeach foreach = (BLangForeach) TreeBuilder.createForeachNode();
+            foreach.pos = queryExpr.pos;
+            foreach.collection = bLangFromClause.collection;
+            types.setForeachTypedBindingPatternType(foreach);
+
+            foreach.variableDefinitionNode = bLangFromClause.variableDefinitionNode;
+            foreach.isDeclaredWithVar = true;
+
+            if (leafForEach != null) {
+                BLangBlockStmt foreachBody = ASTBuilderUtil.createBlockStmt(pos);
+                foreachBody.addStatement(foreach);
+                leafForEach.setBody(foreachBody);
+            } else {
+                parentForEach = foreach;
+            }
+
+            leafForEach = foreach;
+        }
+
+        BLangBlockStmt foreachBody = ASTBuilderUtil.createBlockStmt(pos);
+
+        BType outputArrayType;
+        if (selectClause.expression  != null && selectClause.expression.type != null) {
+            outputArrayType = new BArrayType(selectClause.expression.type);
+        } else {
+            outputArrayType = fromClause.varType;
+        }
+
+        BLangListConstructorExpr emptyArrayExpr = ASTBuilderUtil.createEmptyArrayLiteral(pos,
+                (BArrayType) outputArrayType);
+        BVarSymbol emptyArrayVarSymbol = new BVarSymbol(0, new Name("$outputDataArray$"),
+                this.env.scope.owner.pkgID, outputArrayType, env.scope.owner);
+        BLangSimpleVariable outputArrayVariable =
+                ASTBuilderUtil.createVariable(pos, "$outputDataArray$", outputArrayType,
+                        emptyArrayExpr, emptyArrayVarSymbol);
+
+        // Create temp array variable
+        //      Person[] x = [];
+
+        BLangSimpleVariableDef outputVariableDef =
+                ASTBuilderUtil.createVariableDef(pos, outputArrayVariable);
+        BLangSimpleVarRef outputVarRef = ASTBuilderUtil.createVariableRef(pos, outputArrayVariable.symbol);
+
+        // Create indexed based access expression statement
+        //      x[x.length()] = {
+        //         firstName: person.firstName,
+        //         lastName: person.lastName
+        //      };
+
+        if (selectClause.expression.type == null) {
```

Comment:
this is always not null after typechecker?

Reply:
Fixed.

Reply:
Now, record literals are not type checked hence we need this check now. We can remove this once lang level changes are done.